doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,178 @@
+.. _rewrite_rules:
+
+User-defined rewrite rules
+==========================
+
+.. warning::
+
+   The status of rewrite rules is highly experimental.
+
+   In particular, ill-typed rewrite rules will lead to mistyped expressions,
+   and manipulating these will most often result in anomalies.
+
+
+This section describes the extension of Coq's reduction mechanisms with user-defined rewrite rules,
+as a means to extend definitional equality. It should not be confused with the :ref:`rewrite tactic <rewritingexpressions>`
+or :ref:`setoid rewriting <generalizedrewriting>` which operate on propositional equality and other relations which are defined in Coq.
+
+The use of rewrite rules is guarded behind a flag, which can be enabled
+either by passing ``-allow-rewrite-rules`` to the
+Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.
+
+.. flag:: Allow Rewrite Rules
+
+   This :term:`flag` enables the use of rewrite rules.
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+They stand in-between defined constants and axioms,
+in that they don't always reduce as defined constants do,
+but they may still reduce using the provided rules, unlike axioms.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   As a command, `Symbol` binds an :n:`ident` to a :n:`type` as a symbol.
+
+   .. coqtop:: in
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Rewrite rules
+---------------
+
+  .. insertprodn rewrite_rule rewrite_rule
+
+  .. prodn::
+     rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term
+
+Rewrite rules have two parts simply named left-hand side or pattern and right-hand side.
+When a rule is applied, the term is matched against the pattern,
+subterms aligned with pattern variables are collected
+and substituted into the right-hand side which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
+  :name: Rewrite Rule; Rewrite Rules
+
+   The command declares a named block of rewrite rules.
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+     | ?n ++ 0 ==> ?n
+     | ?n ++ S ?m ==> S (?n ++ ?m)
+     | 0 ++ ?n ==> ?n
+     | S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are a subset of terms which follow a more rigid syntax.
+It is easier to understand them as being a rigid head-pattern, inside of an elimination context:
+
+  .. prodn::
```

Comment:
The problem with adding productions directly in the doc is that there's no checking whatsoever that the grammar matches the code.  I have a queasy feeling that the grammar below may be inaccurate or misleading.

Will the code generate error messages if the user specifies something that's valid in a `pattern` that's not valid for Rewrite?

`elim_context` seems like an arbitrary and fairly meaningless name.  Is there a better name?

It would be good if you make the relationship between the permitted subset of `term` described here with the underlying definitions of term syntax in the `term_match` and `term_forall_or_fun` nonterminals and using similar ordering and naming--to facilitate comparing Rewrite to the general case for patterns.  Also hyperlink to `term_match` and `term_forall_or_fun`.

Is there a better name for `rw_pattern_arg`?

It would be very helpful to add several examples with `coqtop::`, both for simple and more complex cases.

Reply:
Rewrite patterns are parsed from terms, so they don't have a grammar that can be imported from mlg files.

If a term fails to be parsed as a pattern, an error will be raised (there is no "default pattern").

`elim_context` is a term context which represents allowed eliminations for patterns. Naming things is difficult, and I don't have a better name.

I assume the reader is somewhat familiar with the syntax of terms and can pick up on my simplified description of them. I can't simply include `term_match` and `term_forall_or_fun` since they only mention terms and I mention either `rw_pattern_arg` or `elim_context` depending in the occurences. I'm not sure simply mentioning them will help when an intuitive understanding is enough.

Same as `elim_context`, I don't have a better name. As the grammar presents, they stand for the patterns where pattern variables are allowed.

In the examples I give, there is already an instance of every construction except bound variables (which work as any other head pattern), lambdas (which work exactly like products) and primitive projections (which are a much simpler construction than match). I don't understand what cases you feel are missing